homeassistant/components/zha/light.py

```diff
@@ -438,6 +441,28 @@ def async_restore_last_state(self, last_state):
         if "effect" in last_state.attributes:
             self._effect = last_state.attributes["effect"]
 
+    async def async_turn_on(self, **kwargs):
+        """Turn the entity on."""
+        await self._handle_on_off(super().async_turn_on, kwargs)
+
+    async def async_turn_off(self, **kwargs):
+        """Turn the entity off."""
+        await self._handle_on_off(super().async_turn_off, kwargs)
+
+    async def _handle_on_off(self, operation, kwargs):
+        """Perform the on or off operation."""
+        transition = kwargs.get(light.ATTR_TRANSITION)
+        default_transition = DEFAULT_TRANSITION / 10 if DEFAULT_TRANSITION > 1 else 1
+        duration = transition + 0.5 if transition else default_transition + 0.5
+        self._transitioning = True
+        await operation(**kwargs)
+
+        async def do_refresh(_):
+            self._transitioning = False
+            await self.async_get_state()
+
+        async_call_later(self.hass, duration, do_refresh)
+
     async def async_get_state(self):
         """Attempt to retrieve the state from the light."""
         if not self.available:
```

Comment:
Although unlikely, a poll of the light can also happen (with the defined polling interval) while a light is transitioning. Perhaps a check should be added to not poll the light while it's transitioning in ``async_get_state``?
(It's more likely with the non-attribute reporting Hue lights, as they have a polling interval of just a couple of minutes).
This should then also fix https://github.com/home-assistant/core/issues/40300 I think